### ISSUE TITLE
High CPU usage on Kafka Consumer

### ISSUE THREAD

**Issue by wonderedge** [edenhill/librdkafka#3101]
Description
===========
My Kafka cluster have about 80 brokers, while launching kafka consumers. The process CPU usage remains high.

stack info:
```
Thread 376 (LWP 25655):
#0  0x0000ffff97961b04 in poll () from /lib64/libc.so.6
#1  0x0000ffff15b57d10 in rd_kafka_transport_poll (rktrans=0xffff90569c40, tmout=<optimized out>) at rdkafka_transport.c:982
#2  0x0000ffff15b57da4 in rd_kafka_transport_io_serve (rktrans=rktrans@entry=0xffff90569c40, timeout_ms=1000) at rdkafka_transport.c:809
#3  0x0000ffff15b3e87c in rd_kafka_broker_ops_io_serve (rkb=rkb@entry=0xaaaaea9d9350, abs_timeout=abs_timeout@entry=1478412109) at rdkafka_broker.c:3029
#4  0x0000ffff15b3edf8 in rd_kafka_broker_consumer_serve (rkb=rkb@entry=0xaaaaea9d9350, abs_timeout=abs_timeout@entry=1478412109) at rdkafka_broker.c:4544
#5  0x0000ffff15b3fe6c in rd_kafka_broker_serve (rkb=rkb@entry=0xaaaaea9d9350, timeout_ms=<optimized out>, timeout_ms@entry=1000) at rdkafka_broker.c:4646
#6  0x0000ffff15b4027c in rd_kafka_broker_thread_main (arg=0xaaaaea9d9350) at rdkafka_broker.c:4779
#7  0x0000ffff9ba0d7dc in ?? () from /lib64/libpthread.so.0
#8  0x0000ffff9796ba1c in ?? () from /lib64/libc.so.6
Thread 375 (LWP 25654):
#0  0x0000ffff97961b04 in poll () from /lib64/libc.so.6
#1  0x0000ffff15b57d10 in rd_kafka_transport_poll (rktrans=0xaaaae8c0e5a0, tmout=<optimized out>) at rdkafka_transport.c:982
#2  0x0000ffff15b57da4 in rd_kafka_transport_io_serve (rktrans=rktrans@entry=0xaaaae8c0e5a0, timeout_ms=1000) at rdkafka_transport.c:809
#3  0x0000ffff15b3e87c in rd_kafka_broker_ops_io_serve (rkb=rkb@entry=0xaaaaea9d8490, abs_timeout=abs_timeout@entry=1478390109) at rdkafka_broker.c:3029
#4  0x0000ffff15b3edf8 in rd_kafka_broker_consumer_serve (rkb=rkb@entry=0xaaaaea9d8490, abs_timeout=abs_timeout@entry=1478390109) at rdkafka_broker.c:4544
#5  0x0000ffff15b3fe6c in rd_kafka_broker_serve (rkb=rkb@entry=0xaaaaea9d8490, timeout_ms=<optimized out>, timeout_ms@entry=1000) at rdkafka_broker.c:4646
#6  0x0000ffff15b4027c in rd_kafka_broker_thread_main (arg=0xaaaaea9d8490) at rdkafka_broker.c:4779
#7  0x0000ffff9ba0d7dc in ?? () from /lib64/libpthread.so.0
#8  0x0000ffff9796ba1c in ?? () from /lib64/libc.so.6
Thread 374 (LWP 25653):
#0  0x0000ffff97961b04 in poll () from /lib64/libc.so.6
#1  0x0000ffff15b57d10 in rd_kafka_transport_poll (rktrans=0xffff906367e0, tmout=<optimized out>) at rdkafka_transport.c:982
#2  0x0000ffff15b57da4 in rd_kafka_transport_io_serve (rktrans=rktrans@entry=0xffff906367e0, timeout_ms=1000) at rdkafka_transport.c:809
#3  0x0000ffff15b3e87c in rd_kafka_broker_ops_io_serve (rkb=rkb@entry=0xaaaaea9d75d0, abs_timeout=abs_timeout@entry=1477553070) at rdkafka_broker.c:3029
#4  0x0000ffff15b3edf8 in rd_kafka_broker_consumer_serve (rkb=rkb@entry=0xaaaaea9d75d0, abs_timeout=abs_timeout@entry=1477553070) at rdkafka_broker.c:4544
#5  0x0000ffff15b3fe6c in rd_kafka_broker_serve (rkb=rkb@entry=0xaaaaea9d75d0, timeout_ms=<optimized out>, timeout_ms@entry=1000) at rdkafka_broker.c:4646
#6  0x0000ffff15b4041c in rd_kafka_broker_thread_main (arg=0xaaaaea9d75d0) at rdkafka_broker.c:4798
#7  0x0000ffff9ba0d7dc in ?? () from /lib64/libpthread.so.0
#8  0x0000ffff9796ba1c in ?? () from /lib64/libc.so.6
Thread 373 (LWP 25652):
#0  0x0000ffff97961b08 in poll () from /lib64/libc.so.6
#1  0x0000ffff15b57d10 in rd_kafka_transport_poll (rktrans=0xffff8c568f10, tmout=<optimized out>) at rdkafka_transport.c:982
#2  0x0000ffff15b57da4 in rd_kafka_transport_io_serve (rktrans=rktrans@entry=0xffff8c568f10, timeout_ms=474) at rdkafka_transport.c:809
#3  0x0000ffff15b3e87c in rd_kafka_broker_ops_io_serve (rkb=rkb@entry=0xaaaaea9d6710, abs_timeout=abs_timeout@entry=1477992014) at rdkafka_broker.c:3029
#4  0x0000ffff15b3edf8 in rd_kafka_broker_consumer_serve (rkb=rkb@entry=0xaaaaea9d6710, abs_timeout=abs_timeout@entry=1477992014) at rdkafka_broker.c:4544
#5  0x0000ffff15b3fe6c in rd_kafka_broker_serve (rkb=rkb@entry=0xaaaaea9d6710, timeout_ms=<optimized out>, timeout_ms@entry=1000) at rdkafka_broker.c:4646
#6  0x0000ffff15b4027c in rd_kafka_broker_thread_main (arg=0xaaaaea9d6710) at rdkafka_broker.c:4779
#7  0x0000ffff9ba0d7dc in ?? () from /lib64/libpthread.so.0
#8  0x0000ffff9796ba1c in ?? () from /lib64/libc.so.6
Thread 372 (LWP 25651):
#0  0x0000ffff97961b04 in poll () from /lib64/libc.so.6
#1  0x0000ffff15b57d10 in rd_kafka_transport_poll (rktrans=0xaaaaecb650d0, tmout=<optimized out>) at rdkafka_transport.c:982
#2  0x0000ffff15b57da4 in rd_kafka_transport_io_serve (rktrans=rktrans@entry=0xaaaaecb650d0, timeout_ms=1000) at rdkafka_transport.c:809
#3  0x0000ffff15b3e87c in rd_kafka_broker_ops_io_serve (rkb=rkb@entry=0xaaaaea9d5850, abs_timeout=abs_timeout@entry=1478024020) at rdkafka_broker.c:3029
#4  0x0000ffff15b3edf8 in rd_kafka_broker_consumer_serve (rkb=rkb@entry=0xaaaaea9d5850, abs_timeout=abs_timeout@entry=1478024020) at rdkafka_broker.c:4544
#5  0x0000ffff15b3fe6c in rd_kafka_broker_serve (rkb=rkb@entry=0xaaaaea9d5850, timeout_ms=<optimized out>, timeout_ms@entry=1000) at rdkafka_broker.c:4646
#6  0x0000ffff15b4027c in rd_kafka_broker_thread_main (arg=0xaaaaea9d5850) at rdkafka_broker.c:4779
#7  0x0000ffff9ba0d7dc in ?? () from /lib64/libpthread.so.0
#8  0x0000ffff9796ba1c in ?? () from /lib64/libc.so.6
Thread 371 (LWP 25650):
#0  0x0000ffff9ba16ff8 in ?? () from /lib64/libpthread.so.0
#1  0x0000ffff9ba17618 in read () from /lib64/libpthread.so.0
#2  0x0000ffff9863d474 in ?? () from ####/lib/libcrypto.so.1.1
#3  0x0000ffff98635364 in ?? () from ####/lib/libcrypto.so.1.1
#4  0x0000ffff986338b8 in BIO_read () from ####/lib/libcrypto.so.1.1
#5  0x0000ffff984f78a0 in ?? () from ####/lib/libssl.so.1.1
#6  0x0000ffff984fc944 in ?? () from ####/lib/libssl.so.1.1
#7  0x0000ffff984f93cc in ?? () from ####/lib/libssl.so.1.1
#8  0x0000ffff98536370 in ?? () from ####/lib/libssl.so.1.1
#9  0x0000ffff9852c14c in ?? () from ####/lib/libssl.so.1.1
#10 0x0000ffff9851319c in SSL_do_handshake () from ####/lib/libssl.so.1.1
#11 0x0000ffff15bde830 in rd_kafka_transport_ssl_handshake (rktrans=rktrans@entry=0xffff8c5cd390) at rdkafka_ssl.c:578
#12 0x0000ffff15b572ac in rd_kafka_transport_io_event (rktrans=rktrans@entry=0xffff8c5cd390, events=events@entry=4) at rdkafka_transport.c:686
#13 0x0000ffff15b57df0 in rd_kafka_transport_io_serve (rktrans=rktrans@entry=0xffff8c5cd390, timeout_ms=<optimized out>) at rdkafka_transport.c:818
#14 0x0000ffff15b3e87c in rd_kafka_broker_ops_io_serve (rkb=rkb@entry=0xaaaaea9d4990, abs_timeout=abs_timeout@entry=1477808015) at rdkafka_broker.c:3029
#15 0x0000ffff15b3edf8 in rd_kafka_broker_consumer_serve (rkb=rkb@entry=0xaaaaea9d4990, abs_timeout=abs_timeout@entry=1477808015) at rdkafka_broker.c:4544
#16 0x0000ffff15b3fe6c in rd_kafka_broker_serve (rkb=rkb@entry=0xaaaaea9d4990, timeout_ms=<optimized out>, timeout_ms@entry=1000) at rdkafka_broker.c:4646
#17 0x0000ffff15b4027c in rd_kafka_broker_thread_main (arg=0xaaaaea9d4990) at rdkafka_broker.c:4779
#18 0x0000ffff9ba0d7dc in ?? () from /lib64/libpthread.so.0
#19 0x0000ffff9796ba1c in ?? () from /lib64/libc.so.6
Thread 370 (LWP 25649):
#0  0x0000ffff97961b04 in poll () from /lib64/libc.so.6
#1  0x0000ffff15b57d10 in rd_kafka_transport_poll (rktrans=0xffff89be2d10, tmout=<optimized out>) at rdkafka_transport.c:982
#2  0x0000ffff15b57da4 in rd_kafka_transport_io_serve (rktrans=rktrans@entry=0xffff89be2d10, timeout_ms=1000) at rdkafka_transport.c:809
```

How to reproduce
================
This seems occurs while I launched a lot of consumers simultaneously.

Checklist
=========

 - [x] librdkafka version (release number or git tag): `1.3.0`
 - [x] Apache Kafka version: `2.4.1`
 - [x] librdkafka client configuration: `enable.sparse.connections=true; 
reconnect.backoff.jitter.ms=60000; 
auto.commit.interval.ms=60000`
 - [x] Operating system: `Suse 12`
 - [x] Provide logs (with `debug=..` as necessary) from librdkafka
 - [ ] Provide broker log excerpts
 - [ ] Critical issue

**Comment by edenhill** [comment]
How many consumers did you run simultaneously in the same process?
Do you have any actual numbers for CPU usage? Was it steady or did it go up/down after a while?

**Comment by wonderedge** [comment]
Thanks for the quick response. 
Only one consumer in each process, each node have 4 processes, and there are 80 nodes, so there will be 4*80 consumers launched simultaneously.
I have 16 cores in each node, and the consumer process seems take almost all the cpu usage.

**Comment by wonderedge** [comment]
**update**
I checked the kafka broker logs, there were lots of error: `java.net.SocketTimeoutException: Failed to connect within 30000ms`.
so I think this maybe cause by too much connections, the broker was not able to handle in time (or maybe there was some issue with kafka.), and the consumer keeps connect after timeout.
And about the connect to all brokers issue, I believe it was fixed in 1.5.0.
https://github.com/edenhill/librdkafka/commit/1c02f74cb41ed970a695f65393799eadd14267ec
